src/mesh/domain.hpp

```diff
@@ -0,0 +1,181 @@
+//========================================================================================
+// Athena++ astrophysical MHD code
+// Copyright(C) 2014 James M. Stone <jmstone@princeton.edu> and other code contributors
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef MESH_DOMAIN_HPP_
+#define MESH_DOMAIN_HPP_
+
+#include <array>
+
+#include "athena.hpp" 
+#include "athena_arrays.hpp"
+
+namespace parthenon {
+
+  struct IndexRange {
+    int start = 0; /// Starting Index (inclusive)
+    int end = 0; /// Ending Index (inclusive)
+    int ncells() const noexcept { return end-start+1; }
+  };
+  
+  // Assuming we have a block
+  //
+  //  - - - - - - - - - -   ^
+  //  |  |  ghost    |  |   | 
+  //  - - - - - - - - - -   | 
+  //  |  |     ^     |  |   | 
+  //  |  |     |     |  |   
+  //  |  | interior  |  |   entire
+  //  |  |     |     |  |    
+  //  |  |     v     |  |   | 
+  //  - - - - - - - - - -   |
+  //  |  |           |  |   | 
+  //  - - - - - - - - - -   v 
+  //
+  enum IndexShapeType {
+    entire,
+    interior
+  };
```

Comment:
I don't like have C-style enums in the `parthenon` namespace. What if someone wants to use `interior` or `entire` as a variable name in the `parthenon` namespace? I would prefer `enum class`, but then you have to write `IndexShapeType::interior` instead of `interior`. So it might be better to get rid of this enum altogether. See my suggestion below.